syntax = "proto3";

package cerebelum.worker;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";

// WorkerService - Main service for SDK worker communication
service WorkerService {
  // Worker lifecycle management
  rpc Register(RegisterRequest) returns (RegisterResponse);
  rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);
  rpc Unregister(UnregisterRequest) returns (google.protobuf.Empty);

  // Task execution
  rpc PollForTask(PollRequest) returns (Task);
  rpc SubmitResult(TaskResult) returns (Ack);

  // Workflow management
  rpc SubmitBlueprint(Blueprint) returns (BlueprintValidation);
  rpc ExecuteWorkflow(ExecuteRequest) returns (ExecutionHandle);

  // Execution status and control (NEW)
  rpc GetExecutionStatus(GetExecutionStatusRequest) returns (ExecutionStatus);
  rpc ListExecutions(ListExecutionsRequest) returns (ListExecutionsResponse);
  rpc ResumeExecution(ResumeExecutionRequest) returns (ExecutionHandle);
}

// Worker Registration

message RegisterRequest {
  string worker_id = 1;
  string language = 2;  // "kotlin", "typescript", "python"
  repeated string capabilities = 3;  // List of workflow modules this worker can execute
  map<string, string> metadata = 4;  // Additional worker metadata
  string version = 5;  // SDK version
}

message RegisterResponse {
  bool success = 1;
  string message = 2;
  int32 heartbeat_interval_ms = 3;  // How often to send heartbeats (default: 10000)
}

// Heartbeat

message HeartbeatRequest {
  string worker_id = 1;
  WorkerStatus status = 2;
}

enum WorkerStatus {
  WORKER_STATUS_UNSPECIFIED = 0;
  IDLE = 1;
  BUSY = 2;
  DRAINING = 3;  // Graceful shutdown in progress
}

message HeartbeatResponse {
  bool acknowledged = 1;
  repeated string commands = 2;  // Commands from server: "shutdown", "cancel_task", etc.
}

// Unregister

message UnregisterRequest {
  string worker_id = 1;
  string reason = 2;  // "shutdown", "error", etc.
}

// Task Polling

message PollRequest {
  string worker_id = 1;
  int32 timeout_ms = 2;  // Long-polling timeout (max 30000ms)
}

message Task {
  string task_id = 1;
  string execution_id = 2;
  string workflow_module = 3;
  string step_name = 4;
  google.protobuf.Struct step_inputs = 5;  // JSON inputs for the step
  google.protobuf.Struct context = 6;  // Execution context
  google.protobuf.Timestamp created_at = 7;
}

// Task Result Submission

message TaskResult {
  string task_id = 1;
  string execution_id = 2;
  string worker_id = 3;
  TaskStatus status = 4;
  google.protobuf.Struct result = 5;  // JSON result from step execution
  ErrorInfo error = 6;  // Present if status is FAILED
  google.protobuf.Timestamp completed_at = 7;
  SleepRequest sleep_request = 8;  // Present if status is SLEEP
  ApprovalRequest approval_request = 9;  // Present if status is APPROVAL
}

// Sleep Request - step wants to pause workflow
message SleepRequest {
  int64 duration_ms = 1;  // Sleep duration in milliseconds
  google.protobuf.Struct data = 2;  // Data to carry through sleep
}

// Approval Request - step wants human approval
message ApprovalRequest {
  string approval_type = 1;  // "manual", "automated", etc.
  google.protobuf.Struct data = 2;  // Approval data
  int64 timeout_ms = 3;  // Optional timeout
}

enum TaskStatus {
  TASK_STATUS_UNSPECIFIED = 0;
  SUCCESS = 1;
  FAILED = 2;
  TIMEOUT = 3;
  CANCELLED = 4;
  SLEEP = 5;  // Step requested sleep - workflow will pause
  APPROVAL = 6;  // Step requested approval - workflow will wait
}

message ErrorInfo {
  string kind = 1;  // "exception", "timeout", "validation_error", etc.
  string message = 2;
  string stacktrace = 3;
}

message Ack {
  bool success = 1;
  string message = 2;
}

// Blueprint Submission

message Blueprint {
  string workflow_module = 1;  // Fully qualified module name
  string language = 2;  // Source language: "kotlin", "typescript", "python"
  string source_code = 3;  // Optional: source code for debugging
  WorkflowDefinition definition = 4;
  string version = 5;  // Workflow version/hash
}

message WorkflowDefinition {
  repeated Step timeline = 1;
  repeated DivergeRule diverge_rules = 2;
  repeated BranchRule branch_rules = 3;
  map<string, InputDefinition> inputs = 4;
}

message Step {
  string name = 1;
  repeated string depends_on = 2;  // Step dependencies
}

message DivergeRule {
  string from_step = 1;
  repeated PatternMatch patterns = 2;
}

message BranchRule {
  string from_step = 1;
  repeated ConditionBranch branches = 2;
}

message PatternMatch {
  string pattern = 1;  // Pattern to match (e.g., ":timeout", "{:error, :network}")
  string target = 2;  // Target step name
}

message ConditionBranch {
  string condition = 1;  // Boolean expression
  string target = 2;  // Target step name
}

message InputDefinition {
  string type = 1;  // "string", "integer", "boolean", etc.
  bool required = 2;
  google.protobuf.Struct default_value = 3;
}

message BlueprintValidation {
  bool valid = 1;
  repeated string errors = 2;
  repeated string warnings = 3;
  string workflow_hash = 4;  // Hash of the workflow definition
}

// Workflow Execution

message ExecuteRequest {
  string workflow_module = 1;
  google.protobuf.Struct inputs = 2;
  ExecutionOptions options = 3;
}

message ExecutionOptions {
  string execution_id = 1;  // Optional custom execution ID
  string correlation_id = 2;  // For tracing
  repeated string tags = 3;  // Metadata tags
  int32 timeout_ms = 4;  // Execution timeout
}

message ExecutionHandle {
  string execution_id = 1;
  string status = 2;  // "running", "completed", "failed"
  google.protobuf.Timestamp started_at = 3;
}

// Execution Status Query (NEW)

message GetExecutionStatusRequest {
  string execution_id = 1;
}

message ExecutionStatus {
  // Basic info
  string execution_id = 1;
  string workflow_name = 2;
  ExecutionState status = 3;
  google.protobuf.Timestamp started_at = 4;
  google.protobuf.Timestamp completed_at = 5;  // Present if completed/failed

  // Progress info
  int32 current_step_index = 6;  // 0-indexed
  int32 total_steps = 7;
  string current_step_name = 8;  // Empty if completed

  // Completed steps
  repeated StepStatus completed_steps = 9;

  // Inputs and outputs
  google.protobuf.Struct inputs = 10;
  map<string, google.protobuf.Struct> step_outputs = 11;

  // Error if failed
  ErrorInfo error = 12;

  // Sleep info (if sleeping)
  SleepInfo sleep_info = 13;

  // Approval info (if waiting for approval)
  ApprovalInfo approval_info = 14;
}

message StepStatus {
  string step_name = 1;
  int32 step_index = 2;
  string status = 3;  // "completed", "failed", "running"
  google.protobuf.Timestamp started_at = 4;
  google.protobuf.Timestamp completed_at = 5;
  int32 duration_seconds = 6;
  google.protobuf.Struct output = 7;
  ErrorInfo error = 8;  // Present if step failed
}

message SleepInfo {
  int64 duration_ms = 1;  // Total sleep duration
  google.protobuf.Timestamp sleep_started_at = 2;
  int64 remaining_ms = 3;  // Remaining time
  google.protobuf.Struct data = 4;  // Data carried through sleep
}

message ApprovalInfo {
  string approval_type = 1;
  google.protobuf.Struct data = 2;
  int64 timeout_ms = 3;
  google.protobuf.Timestamp requested_at = 4;
  int64 remaining_timeout_ms = 5;
}

enum ExecutionState {
  EXECUTION_STATE_UNSPECIFIED = 0;
  EXECUTION_RUNNING = 1;
  EXECUTION_COMPLETED = 2;
  EXECUTION_FAILED = 3;
  EXECUTION_SLEEPING = 4;
  EXECUTION_WAITING_FOR_APPROVAL = 5;
  EXECUTION_PAUSED = 6;
}

// List Executions (NEW)

message ListExecutionsRequest {
  optional string workflow_name = 1;  // Filter by workflow name
  optional ExecutionState status = 2;  // Filter by status
  int32 limit = 3;  // Default: 50, max: 100
  int32 offset = 4;  // For pagination
}

message ListExecutionsResponse {
  repeated ExecutionStatus executions = 1;
  int32 total_count = 2;
  bool has_more = 3;
}

// Resume Execution (NEW)

message ResumeExecutionRequest {
  string execution_id = 1;
  bool skip_completed_steps = 2;  // Default: true
}
